### ISSUE TITLE
3.1.0: Nine test_functional and one test_package fails

### ISSUE THREAD

**Issue by mtelka** [pylint-dev/pylint#9477]
### Bug description

I'm running tests for `pylint 3.1.0` and I found that the following tests fails:

```
=========================== short test summary info ============================
FAILED tests/test_functional.py::test_functional[inconsistent_returns] - Asse...
FAILED tests/test_functional.py::test_functional[invalid_all_format_valid_5]
FAILED tests/test_functional.py::test_functional[invalid_all_format_valid_6]
FAILED tests/test_functional.py::test_functional[regression_4680] - Assertion...
FAILED tests/test_functional.py::test_functional[unreachable] - AssertionErro...
FAILED tests/test_functional.py::test_functional[undefined_variable_typing]
FAILED tests/test_functional.py::test_functional[unused_import_assigned_to]
FAILED tests/test_functional.py::test_functional[use_yield_from] - AssertionE...
FAILED tests/test_functional.py::test_functional[used_before_assignment_else_return]
FAILED tests/test_regr.py::test_package[file_names1-<lambda>] - assert False
```

### Command used

```shell
tox --current-env --no-provision --recreate -e py39
```

### Pylint output

```python
self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x7fff95c24a60>

    def runTest(self) -> None:
>       self._runTest()
E       AssertionError: Wrong message(s) raised for "inconsistent_returns.py":
E
E       Unexpected in testdata:
E        111: inconsistent-return-statements
E
E       Actual pylint output for this file:
E       OutputLine(symbol='inconsistent-return-statements', lineno=111, column=0, end_lineno=111, end_column=12, object='bug_1771', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=160, column=0, end_lineno=160, end_column=29, object='explicit_implicit_returns', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=164, column=0, end_lineno=164, end_column=26, object='empty_explicit_returns', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=169, column=0, end_lineno=169, end_column=30, object='explicit_implicit_returns2', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=177, column=0, end_lineno=177, end_column=30, object='explicit_implicit_returns3', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=185, column=0, end_lineno=185, end_column=41, object='returns_missing_in_catched_exceptions', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=195, column=0, end_lineno=195, end_column=16, object='complex_func', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=203, column=4, end_lineno=203, end_column=36, object='inconsistent_returns_in_nested_function.not_consistent_returns_inner', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=211, column=0, end_lineno=211, end_column=28, object='bug_1771_counter_example', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=229, column=0, end_lineno=229, end_column=28, object='bug_1772_counter_example', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=237, column=0, end_lineno=237, end_column=47, object='bug_1794_inner_func_in_if_counter_example_1', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=246, column=0, end_lineno=246, end_column=47, object='bug_1794_inner_func_in_if_counter_example_2', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=255, column=0, end_lineno=255, end_column=47, object='bug_1794_inner_func_in_if_counter_example_3', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=262, column=8, end_lineno=262, end_column=19, object='bug_1794_inner_func_in_if_counter_example_3._inner2', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=267, column=0, end_lineno=267, end_column=12, object='bug_3468', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=277, column=0, end_lineno=277, end_column=20, object='bug_3468_variant', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=322, column=0, end_lineno=322, end_column=21, object='bug_pylint_3873_1', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=349, column=0, end_lineno=349, end_column=25, object='bug_pylint_4019_wrong', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=360, column=4, end_lineno=360, end_column=22, object='A.get_the_answer', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')
E       OutputLine(symbol='inconsistent-return-statements', lineno=367, column=0, end_lineno=367, end_column=18, object='bug_1772_break', msg='Either all return statements in a function should return an expression, or none of them should.', confidence='UNDEFINED')

pylint/testutils/lint_module_test.py:147: AssertionError
_________________ test_functional[invalid_all_format_valid_5] __________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x7fff94952820>

    def runTest(self) -> None:
>       self._runTest()
E       AssertionError: Wrong message(s) raised for "invalid_all_format_valid_5.py":
E
E       Unexpected in testdata:
E          5: no-name-in-module
E
E       Actual pylint output for this file:
E       OutputLine(symbol='no-name-in-module', lineno=5, column=0, end_lineno=5, end_column=19, object='', msg="No name 'bar' in module 'foo'", confidence='UNDEFINED')

pylint/testutils/lint_module_test.py:147: AssertionError
_________________ test_functional[invalid_all_format_valid_6] __________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x7fff948abf10>

    def runTest(self) -> None:
>       self._runTest()
E       AssertionError: Wrong message(s) raised for "invalid_all_format_valid_6.py":
E
E       Unexpected in testdata:
E          5: no-name-in-module
E
E       Actual pylint output for this file:
E       OutputLine(symbol='no-name-in-module', lineno=5, column=0, end_lineno=5, end_column=19, object='', msg="No name 'bar' in module 'foo'", confidence='UNDEFINED')

pylint/testutils/lint_module_test.py:147: AssertionError
_______________________ test_functional[regression_4680] _______________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x7fff901aea00>

    def runTest(self) -> None:
>       self._runTest()
E       AssertionError: Wrong message(s) raised for "regression_4680.py":
E
E       Unexpected in testdata:
E          3: no-name-in-module
E         18: c-extension-no-member
E
E       Actual pylint output for this file:
E       OutputLine(symbol='import-error', lineno=3, column=0, end_lineno=3, end_column=14, object='', msg="Unable to import 'foo.sub'", confidence='UNDEFINED')
E       OutputLine(symbol='no-name-in-module', lineno=3, column=0, end_lineno=3, end_column=14, object='', msg="No name 'sub' in module 'foo'", confidence='UNDEFINED')
E       OutputLine(symbol='undefined-variable', lineno=10, column=0, end_lineno=10, end_column=15, object='FailedTwo', msg="Undefined variable 'ab'", confidence='UNDEFINED')
E       OutputLine(symbol='c-extension-no-member', lineno=18, column=7, end_lineno=18, end_column=14, object='', msg="Module 'foo' has no 'sub' member, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects.", confidence='INFERENCE')

pylint/testutils/lint_module_test.py:147: AssertionError
_________________________ test_functional[unreachable] _________________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x7fff8c2b02e0>

    def runTest(self) -> None:
>       self._runTest()
E       AssertionError: Wrong message(s) raised for "unreachable.py":
E
E       Expected in testdata:
E         42: unreachable
E         80: unreachable
E
E       Actual pylint output for this file:
E       OutputLine(symbol='unreachable', lineno=10, column=4, end_lineno=10, end_column=24, object='func1', msg='Unreachable code', confidence='HIGH')
E       OutputLine(symbol='unreachable', lineno=15, column=8, end_lineno=15, end_column=28, object='func2', msg='Unreachable code', confidence='HIGH')
E       OutputLine(symbol='unreachable', lineno=21, column=8, end_lineno=21, end_column=28, object='func3', msg='Unreachable code', confidence='HIGH')
E       OutputLine(symbol='unreachable', lineno=25, column=4, end_lineno=25, end_column=16, object='func4', msg='Unreachable code', confidence='HIGH')
E       OutputLine(symbol='unreachable', lineno=38, column=4, end_lineno=38, end_column=24, object='func6', msg='Unreachable code', confidence='HIGH')
E       OutputLine(symbol='unreachable', lineno=64, column=4, end_lineno=64, end_column=15, object='func9', msg='Unreachable code', confidence='INFERENCE')
E       OutputLine(symbol='unreachable', lineno=69, column=4, end_lineno=69, end_column=15, object='func10', msg='Unreachable code', confidence='INFERENCE')
E       OutputLine(symbol='unreachable', lineno=74, column=4, end_lineno=74, end_column=15, object='func11', msg='Unreachable code', confidence='INFERENCE')

pylint/testutils/lint_module_test.py:147: AssertionError
__________________ test_functional[undefined_variable_typing] __________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x7fff8b3ee520>

    def runTest(self) -> None:
>       self._runTest()
E       AssertionError: Wrong message(s) raised for "undefined_variable_typing.py":
E
E       Unexpected in testdata:
E         10: no-name-in-module
E
E       Actual pylint output for this file:
E       OutputLine(symbol='no-name-in-module', lineno=10, column=0, end_lineno=10, end_column=24, object='', msg="No name 'Bar' in module 'foo'", confidence='UNDEFINED')
E       OutputLine(symbol='no-name-in-module', lineno=10, column=0, end_lineno=10, end_column=24, object='', msg="No name 'Boo' in module 'foo'", confidence='UNDEFINED')

pylint/testutils/lint_module_test.py:147: AssertionError
__________________ test_functional[unused_import_assigned_to] __________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x7fff8a1ef430>

    def runTest(self) -> None:
>       self._runTest()
E       AssertionError: Wrong message(s) raised for "unused_import_assigned_to.py":
E
E       Unexpected in testdata:
E         25: c-extension-no-member
E
E       Actual pylint output for this file:
E       OutputLine(symbol='c-extension-no-member', lineno=25, column=10, end_lineno=25, end_column=17, object='bar', msg="Module 'foo' has no 'baz' member, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects.", confidence='INFERENCE')

pylint/testutils/lint_module_test.py:147: AssertionError
_______________________ test_functional[use_yield_from] ________________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x7fff88dcbf40>

    def runTest(self) -> None:
>       self._runTest()
E       AssertionError: Wrong message(s) raised for "use_yield_from.py":
E
E       Unexpected in testdata:
E          3: no-name-in-module
E         28: no-member
E         33: no-member
E
E       Actual pylint output for this file:
E       OutputLine(symbol='no-name-in-module', lineno=3, column=0, end_lineno=3, end_column=34, object='', msg="No name 'shazam' in module 'magic'", confidence='UNDEFINED')
E       OutputLine(symbol='no-name-in-module', lineno=3, column=0, end_lineno=3, end_column=34, object='', msg="No name 'turbogen' in module 'magic'", confidence='UNDEFINED')
E       OutputLine(symbol='use-yield-from', lineno=8, column=4, end_lineno=9, end_column=18, object='bad', msg="Use 'yield from' directly instead of yielding each element one by one", confidence='HIGH')
E       OutputLine(symbol='use-yield-from', lineno=13, column=4, end_lineno=14, end_column=18, object='out_of_names', msg="Use 'yield from' directly instead of yielding each element one by one", confidence='HIGH')
E       OutputLine(symbol='no-member', lineno=28, column=16, end_lineno=28, end_column=27, object='yield_attr', msg="Module 'factory' has no 'gen' member", confidence='INFERENCE')
E       OutputLine(symbol='use-yield-from', lineno=28, column=4, end_lineno=29, end_column=18, object='yield_attr', msg="Use 'yield from' directly instead of yielding each element one by one", confidence='HIGH')
E       OutputLine(symbol='no-member', lineno=33, column=16, end_lineno=33, end_column=28, object='yield_attr_nested', msg="Module 'factory' has no 'kiwi' member", confidence='INFERENCE')
E       OutputLine(symbol='use-yield-from', lineno=33, column=4, end_lineno=34, end_column=18, object='yield_attr_nested', msg="Use 'yield from' directly instead of yielding each element one by one", confidence='HIGH')
E       OutputLine(symbol='use-yield-from', lineno=38, column=4, end_lineno=39, end_column=18, object='yield_expr', msg="Use 'yield from' directly instead of yielding each element one by one", confidence='HIGH')

pylint/testutils/lint_module_test.py:147: AssertionError
_____________ test_functional[used_before_assignment_else_return] ______________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x7fff88a4cca0>

    def runTest(self) -> None:
>       self._runTest()
E       AssertionError: Wrong message(s) raised for "used_before_assignment_else_return.py":
E
E       Unexpected in testdata:
E         71: used-before-assignment
E
E       Actual pylint output for this file:
E       OutputLine(symbol='used-before-assignment', lineno=26, column=14, end_lineno=26, end_column=19, object='invalid', msg="Using variable 'error' before assignment", confidence='CONTROL_FLOW')
E       OutputLine(symbol='used-before-assignment', lineno=39, column=14, end_lineno=39, end_column=19, object='invalid_2', msg="Using variable 'error' before assignment", confidence='CONTROL_FLOW')
E       OutputLine(symbol='used-before-assignment', lineno=51, column=14, end_lineno=51, end_column=19, object='invalid_3', msg="Using variable 'error' before assignment", confidence='CONTROL_FLOW')
E       OutputLine(symbol='used-before-assignment', lineno=61, column=14, end_lineno=61, end_column=19, object='invalid_4', msg="Using variable 'error' before assignment", confidence='CONTROL_FLOW')
E       OutputLine(symbol='used-before-assignment', lineno=71, column=7, end_lineno=71, end_column=18, object='valid_exit', msg="Using variable 'lint_result' before assignment", confidence='CONTROL_FLOW')

pylint/testutils/lint_module_test.py:147: AssertionError
______________________ test_package[file_names1-<lambda>] ______________________

finalize_linter = Checker 'main' (responsible for 'F0001', 'F0002', 'F0010', 'F0011', 'I0001', 'I0010', 'I0011', 'I0013', 'I0020', 'I0021', 'I0022', 'E0001', 'E0011', 'W0012', 'R0022', 'E0013', 'E0014', 'E0015')
file_names = ['precedence_test']
check = <function Equals.<locals>.<lambda> at 0x7fffaa83d3a0>

    @pytest.mark.parametrize(
        "file_names, check",
        [
            (["package.__init__"], Equals("")),
            (["precedence_test"], Equals("")),
            (["import_package_subpackage_module"], Equals("")),
            (["pylint.checkers.__init__"], lambda x: "__path__" not in x),
            ([join(REGR_DATA, "classdoc_usage.py")], Equals("")),
            ([join(REGR_DATA, "module_global.py")], Equals("")),
            ([join(REGR_DATA, "decimal_inference.py")], Equals("")),
            ([join(REGR_DATA, "absimp", "string.py")], Equals("")),
            ([join(REGR_DATA, "bad_package")], lambda x: "Unused import missing" in x),
        ],
    )
    def test_package(
        finalize_linter: PyLinter, file_names: list[str], check: Callable[[str], bool]
    ) -> None:
        finalize_linter.check(file_names)
        finalize_linter.reporter = cast(  # Due to fixture
            testutils.GenericTestReporter, finalize_linter.reporter
        )
        got = finalize_linter.reporter.finalize().strip()
>       assert check(got)
E       assert False
E        +  where False = <function Equals.<locals>.<lambda> at 0x7fffaa83d3a0>("E: 21: Module 'package.AudioTime' has no 'DECIMAL' member")

tests/test_regr.py:77: AssertionError
```

### Expected behavior

All tests pass.

### Pylint version

```shell
pylint 3.1.0
astroid 3.1.0
Python 3.9.18
```

### OS / Environment

OpenIndiana

**Comment by mtelka** [comment]
It looks like part of test failures is caused by `sys.exit()`.  Apparently, `pylint` thinks the `sys.exit()` call could return, but tests are written with expectation that `sys.exit()` cannot return.

Any suggestion what might be wrong here?

Thank you.

**Comment by Pierre-Sassoulas** [comment]
I don't think any active maintainers can test on solaris. One possibility to troubleshoot would be to look at the generated ast (from the builtin python ast module) then the reconstructed astroid ast for the failing files in order to narrow down the issue to pylint, astroid or python for solaris.

**Comment by mtelka** [comment]
The problem is that I ran tests with pytest 8.  I see similar failures in PR #9396 checks.

**Comment by Pierre-Sassoulas** [comment]
Thanks for the investigation, and indirectly helping with the pytest upgrade :)
